internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,328 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package orchestrator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/copilot-cli/internal/pkg/docker/dockerengine"
+)
+
+// Orchestrator manages running a Task. Only a single Task
+// can be running at a time for a given Orchestrator.
+type Orchestrator struct {
+	idPrefix   string
+	logOptions logOptionsFunc
+
+	curTask   Task
+	curTaskID atomic.Int32
+	runErrs   chan error
+	stopped   chan struct{}
+	wg        *sync.WaitGroup
+	actions   chan action
+	stopOnce  *sync.Once
+
+	docker DockerEngine
+}
+
+type action interface {
+	Do(o *Orchestrator) error
+}
+
+type logOptionsFunc func(name string, ctr ContainerDefinition) dockerengine.RunLogOptions
+
+// DockerEngine is used by Orchestrator to manage containers.
+type DockerEngine interface {
+	Run(context.Context, *dockerengine.RunOptions) error
+	IsContainerRunning(context.Context, string) (bool, error)
+	Stop(context.Context, string) error
+}
+
+const (
+	stoppedTaskID  = -1
```

Comment:
I think my only nit is the naming of this variable, it reads like "this is the ID of the most recently stopped task", while it is actually a special ID that signals that the `Stop()` is already called. Perhaps something like `stopped`... or use a separate boolean variable to convey this message.